cli/cluster/etcd.go

```diff
@@ -0,0 +1,247 @@
+		}
+
+		if opts.CaPath != "" {
+			var err error
+			tlsConfig.RootCAs, err = loadRootCA(opts.CaPath)
+			if err != nil {
+				return nil, fmt.Errorf("fail to load CA directory: %w", err)
+			}
+		}
+
+		if opts.SkipHostVerify {
+			tlsConfig.InsecureSkipVerify = true
+		}
+	}
+
+	return clientv3.New(clientv3.Config{
+		Endpoints:   opts.Endpoints,
+		DialTimeout: opts.Timeout,
+		Username:    opts.Username,
+		Password:    opts.Password,
+		TLS:         tlsConfig,
+	})
+}
+
+// EtcdCollector collects data from a etcd connection.
+type EtcdCollector struct {
+	getter  EtcdGetter
+	prefix  string
+	timeout time.Duration
+}
+
+// EtcdGetter is the interface that wraps get from etcd method.
+type EtcdGetter interface {
+	Get(ctx context.Context, key string, opts ...clientv3.OpOption) (*clientv3.GetResponse, error)
+}
+
+// NewEtcdCollector creates a new collector for etcd from the path.
+func NewEtcdCollector(getter EtcdGetter, prefix string, timeout time.Duration) EtcdCollector {
+	return EtcdCollector{
+		getter:  getter,
+		prefix:  prefix,
+		timeout: timeout,
+	}
+}
+
+// Collect collects a configuration from the specified path with the specified
+// timeout.
+func (collector EtcdCollector) Collect() (*Config, error) {
+	prefix := strings.TrimRight(collector.prefix, "/")
+	key := fmt.Sprintf("%s/%s/", prefix, "config")
+	ctx := context.Background()
+	if collector.timeout != 0 {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, collector.timeout)
+		defer cancel()
+	}
+
+	resp, err := collector.getter.Get(ctx, key, clientv3.WithPrefix())
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch data from etcd: %w", err)
+	}
+
+	cconfig := NewConfig()
+	for _, kv := range resp.Kvs {
+		if config, err := NewYamlCollector(kv.Value).Collect(); err != nil {
+			fmtErr := "failed to decode etcd config for key %q: %w"
+			return nil, fmt.Errorf(fmtErr, string(kv.Key), err)
+		} else {
+			cconfig.Merge(config)
+		}
+	}
+
+	return cconfig, nil
+}
+
+// loadRootCA and the code below is a copy-paste from file
+// crypto/x509/root_unix.go , go 1.20.
+func loadRootCA(path string) (*x509.CertPool, error) {
+	roots := x509.NewCertPool()
+
+	fis, err := readUniqueDirectoryEntries(path)
+	if err != nil {
+		return nil, err
+	}
+
+	rootsLen := 0
+	for _, fi := range fis {
+		data, err := os.ReadFile(path + "/" + fi.Name())
+		if err == nil {
+			rootsLen++
+			roots.AppendCertsFromPEM(data)
+		}
+	}
+
+	return roots, nil
+}
+
+// readUniqueDirectoryEntries is like os.ReadDir but omits
+// symlinks that point within the directory.
+func readUniqueDirectoryEntries(dir string) ([]fs.DirEntry, error) {
+	files, err := os.ReadDir(dir)
+	if err != nil {
+		return nil, err
+	}
+
+	uniq := files[:0]
+	for _, f := range files {
+		if !isSameDirSymlink(f, dir) {
+			uniq = append(uniq, f)
+		}
+	}
+
+	return uniq, nil
+}
+
+// isSameDirSymlink reports whether fi in dir is a symlink with a
```

Comment:
fi?

Reply:
Yeah...

https://cs.opensource.google/go/go/+/refs/tags/go1.21.0:src/crypto/x509/root_unix.go;l=100

Reply:
okay...

Reply:
I've added a links to an original implementation for a copy-paste code.